pkg/model/workflow.go

```diff
@@ -173,65 +207,84 @@ func (j *Job) Environment() map[string]string {
 	return nil
 }
 
 // GetMatrixes returns the matrix cross product
+// It skips includes and hard fails excludes for non-existing keys
+// nolint:gocyclo
 func (j *Job) GetMatrixes() []map[string]interface{} {
 	matrixes := make([]map[string]interface{}, 0)
 	if j.Strategy != nil {
-		m := j.Matrix()
-		includes := make([]map[string]interface{}, 0)
-		for _, v := range m["include"] {
-			switch t := v.(type) {
-			case []interface{}:
-				for _, i := range t {
-					includes = append(includes, i.(map[string]interface{}))
+		j.Strategy.FailFast = j.Strategy.GetFailFast()
+		j.Strategy.MaxParallel = j.Strategy.GetMaxParallel()
+
+		if m := j.Matrix(); m != nil {
+			includes := make([]map[string]interface{}, 0)
+			for _, v := range m["include"] {
+				switch t := v.(type) {
+				case []interface{}:
+					for _, i := range t {
+						i := i.(map[string]interface{})
+						for k := range i {
+							if _, ok := m[k]; ok {
+								includes = append(includes, i)
+								break
+							}
+						}
+					}
+				case interface{}:
+					v := v.(map[string]interface{})
+					for k := range v {
+						if _, ok := m[k]; ok {
+							includes = append(includes, v)
+							break
+						}
+					}
 				}
-			case interface{}:
-				includes = append(includes, v.(map[string]interface{}))
 			}
-		}
-		delete(m, "include")
-
-		excludes := make([]map[string]interface{}, 0)
-		for _, v := range m["exclude"] {
-			excludes = append(excludes, v.(map[string]interface{}))
-		}
-		delete(m, "exclude")
-
-		matrixProduct := common.CartesianProduct(m)
-
-	MATRIX:
-		for _, matrix := range matrixProduct {
-			for _, exclude := range excludes {
-				if commonKeysMatch(matrix, exclude) {
-					log.Debugf("Skipping matrix '%v' due to exclude '%v'", matrix, exclude)
-					continue MATRIX
+			delete(m, "include")
+
+			excludes := make([]map[string]interface{}, 0)
+			for _, e := range m["exclude"] {
+				e := e.(map[string]interface{})
+				for k := range e {
+					if _, ok := m[k]; ok {
+						excludes = append(excludes, e)
+					} else {
+						// We fail completely here because that's what GitHub does for non-existing matrix keys, fail on exclude, silent skip on include
+						log.Fatalf("The workflow is not valid. Matrix exclude key '%s' does not match any key within the matrix", k)
+					}
 				}
 			}
-			matrixes = append(matrixes, matrix)
-		}
-		for _, include := range includes {
-			log.Debugf("Adding include '%v'", include)
-			matrixes = append(matrixes, include)
+			delete(m, "exclude")
+
+			matrixProduct := common.CartesianProduct(m)
+		MATRIX:
+			for _, matrix := range matrixProduct {
+				for _, exclude := range excludes {
+					if commonKeysMatch(matrix, exclude) {
+						log.Debugf("Skipping matrix '%v' due to exclude '%v'", matrix, exclude)
+						continue MATRIX
+					}
+				}
+				matrixes = append(matrixes, matrix)
+			}
+			for _, include := range includes {
+				log.Debugf("Adding include '%v'", include)
+				matrixes = append(matrixes, include)
+			}
```

Comment:
Nothing which blocks this PR,
I think you aren't really aware of the behavior of the matrix include statement, (this is not introduced in this PR, but not fixed either)

Example Workflow run:
https://github.com/actions/runner/actions/runs/1082604590/workflow

This workflow demonstrates that includes needs to be handeled differently, if your matrix cross product is non empty and you have includes.

In [my c# implementation](https://github.com/ChristopherHX/runner.server/blob/5d55db086c97e9bc45e97f18fadaf4043588fdf0/src/Runner.Server/Controllers/MessageController.cs#L890-L914) of this,
- Create the cross product (this seems to work in act with this PR)
- apply excludes (this seems to work in act with this PR)
- Merge the cross product if all member which are both in the cross product and include mapping are equal (added members or matrix entries to the cross product via include are not considered), otherwise do it like this code does.

Well this might be not perfect either, but this seems for me to be more like github actions.